diff --git a/dep/g3dlite/source/Matrix4.cpp b/dep/g3dlite/source/Matrix4.cpp
index 2ce14f6..fbc918f 100644
--- a/dep/g3dlite/source/Matrix4.cpp
+++ b/dep/g3dlite/source/Matrix4.cpp
@@ -382,7 +382,7 @@ bool Matrix4::operator==(const Matrix4& other) const {
     // If the bit patterns are identical, they must be
     // the same matrix.  If not, they *might* still have
     // equal elements due to floating point weirdness.
-    if (memcmp(this, &other, sizeof(Matrix4) == 0)) {
+    if (memcmp(this, &other, sizeof(Matrix4)) == 0) {
         return true;
     } 
 
