diff --git a/dep/g3dlite/include/G3D/platform.h b/dep/g3dlite/include/G3D/platform.h
index 17e3bf2..439495a 100644
--- a/dep/g3dlite/include/G3D/platform.h
+++ b/dep/g3dlite/include/G3D/platform.h
@@ -364,13 +364,18 @@ int WINAPI WinMain(HINSTANCE hInst, HINSTANCE hPrev, LPSTR szCmdLine, int sw) {\
 #endif
 #if (!defined(_LIBCPP_VERSION) && defined(__APPLE__)) || (!defined(_LIBCPP_VERSION) && defined(__linux__))
 #   include <tr1/memory>
+#else
+#   include <memory>
+#endif
+
+namespace G3D {
+#if (!defined(_LIBCPP_VERSION) && defined(__APPLE__)) || (!defined(_LIBCPP_VERSION) && defined(__linux__))
     using std::tr1::shared_ptr;
     using std::tr1::weak_ptr;
     using std::tr1::dynamic_pointer_cast;
     using std::tr1::static_pointer_cast;
     using std::tr1::enable_shared_from_this;
 #else
-#   include <memory>
     using std::shared_ptr;
     using std::weak_ptr;
     using std::dynamic_pointer_cast;
@@ -378,7 +383,6 @@ int WINAPI WinMain(HINSTANCE hInst, HINSTANCE hPrev, LPSTR szCmdLine, int sw) {\
     using std::enable_shared_from_this;
 #endif
 
-namespace G3D {
     /** Options for initG3D and initGLG3D. */
     class G3DSpecification {
     public:
