diff --git a/dep/SFMT/SFMT.h b/dep/SFMT/SFMT.h
index 3d15d65..ccf21ce 100644
--- a/dep/SFMT/SFMT.h
+++ b/dep/SFMT/SFMT.h
@@ -173,7 +173,8 @@ public:
         uint32_t statesize = SFMT_N*4;      // Size of state vector
 
         // Fill state vector with random numbers from seed
-        ((uint32_t*)state)[0] = y;
+        uint32_t* s = (uint32_t*)&state;
+        s[0] = y;
         const uint32_t factor = 1812433253U;// Multiplication factor
 
         for (i = 1; i < statesize; i++) {
